pkg/trace/agent/agent.go

```diff
@@ -274,30 +270,25 @@ func (a *Agent) Process(t pb.Trace) {
 		return
 	}
 	// Run both full trace sampling and transaction extraction in another goroutine.
-	go func(pt ProcessedTrace) {
-		defer watchdog.LogOnPanic()
-		defer timing.Since("datadog.trace_agent.internal.sample_ms", time.Now())
+	var sampledSpans writer.SampledSpans
 
-		var sampledSpans writer.SampledSpans
-
-		sampled, rate := a.sample(pt)
-		if sampled {
-			pt.Sampled = sampled
-			sampler.AddGlobalRate(pt.Root, rate)
-			sampledSpans.Trace = pt.Trace
-		}
+	sampled, rate := a.sample(pt)
+	if sampled {
+		pt.Sampled = sampled
+		sampler.AddGlobalRate(pt.Root, rate)
+		sampledSpans.Trace = pt.Trace
+	}
 
-		// NOTE: Events can be processed on non-sampled traces.
-		events, numExtracted := a.EventProcessor.Process(pt.Root, pt.Trace)
-		sampledSpans.Events = events
+	// NOTE: Events can be processed on non-sampled traces.
+	events, numExtracted := a.EventProcessor.Process(pt.Root, pt.Trace)
```

Comment:
This block of code immediately following the trace sampling is a little confusing.  To help with readability, I'd suggest moving the event sampling into it's own function.  So you have:

```go
	sampled, rate := a.sample(pt)

 	if sampled {
		sampled, rate := a.sample(pt)			pt.Sampled = sampled
		if sampled {			sampler.AddGlobalRate(pt.Root, rate)
			pt.Sampled = sampled			sampledSpans.Trace = pt.Trace
			sampler.AddGlobalRate(pt.Root, rate)		}

   performEventProcessing(pt)
```

Or something like that.

Reply:
Your formatting went a little bit off road, but if I read correctly you're suggesting putting two lines of code into a separate function? I'm not sure if that improves things tbh.

Reply:
Move:
```go
	events, numExtracted := a.EventProcessor.Process(pt.Root, pt.Trace)
	sampledSpans.Events = events
	atomic.AddInt64(&ts.EventsExtracted, int64(numExtracted))
	atomic.AddInt64(&ts.EventsSampled, int64(len(sampledSpans.Events)))
```

It's 4 lines, but it's about making the intent of the code more explicit.  As structured, I need to read code and infer what this is doing and why it's different from the previous lines that are sampling traces.  If you move that to a function then the code becomes a single line that explicitly states what it's doing and I no longer have to infer why it's different.

Reply:
I tried it, I'm not sure if I like this better:
```diff
diff --git a/pkg/trace/agent/agent.go b/pkg/trace/agent/agent.go
index 36f61f19..203ac803 100644
--- a/pkg/trace/agent/agent.go
+++ b/pkg/trace/agent/agent.go
@@ -279,18 +279,20 @@ func (a *Agent) Process(t pb.Trace) {
 		sampledSpans.Trace = pt.Trace
 	}
 
-	// NOTE: Events can be processed on non-sampled traces.
-	events, numExtracted := a.EventProcessor.Process(pt.Root, pt.Trace)
-	sampledSpans.Events = events
-
-	atomic.AddInt64(&ts.EventsExtracted, int64(numExtracted))
-	atomic.AddInt64(&ts.EventsSampled, int64(len(sampledSpans.Events)))
+	sampledSpans.Events = a.sampleEvents(ts, pt.Root, pt.Trace)
 
 	if !sampledSpans.Empty() {
 		a.spansOut <- &sampledSpans
 	}
 }
 
+func (a *Agent) sampleEvents(ts *info.TagStats, root *pb.Span, trace pb.Trace) []*pb.Span {
+	events, numExtracted := a.EventProcessor.Process(root, trace)
+	atomic.AddInt64(&ts.EventsExtracted, int64(numExtracted))
+	atomic.AddInt64(&ts.EventsSampled, int64(len(events)))
+	return events
+}
+
 func (a *Agent) sample(pt ProcessedTrace) (sampled bool, rate float64) {
 	var sampledPriority, sampledScore bool
 	var ratePriority, rateScore float64
```
The signature of the function is a bit gross.

Reply:
Let me know if you were thinking otherwise or if you still feel this is an improvement. I feel that seeing a new function signature, as opposed to the actual code, is not necessarily better. Because even if the function name is hinting at what it's doing, you never really know until you read the code of it, so, now you need to also scroll around to see that.

I don't feel strongly about it, I'm just debating. Never the less, this whole `Process` function is a bit large and could use some love.